docs/reference/api/platform/FileHandle.md

```diff
@@ -0,0 +1,274 @@
+    
+    // Do do:
+    printf("Hello!\n");  // assume platform.stdio-convert-newlines is true
+
+
+Beyond the target-specific override, an application can override the target's default behaviour itself by providing `mbed::mbed_override_console`. 
+
+Alternatively, an application could use the standard C `freopen` function to redirect `stdout` to a named file or device while running. However there is no `fdreopen` analogue to redirect to an unnamed device by file descriptor or `FileHandle` pointer. 
+
+### Polling and non-blocking
+
+By default `FileHandle`s conventionally block until a `read` or `write` operation completes. This is the only behaviour supported by normal `File`s, and is expected by the C library's `stdio` functions.
+
+Device-type `FileHandle`s such as `UARTSerial` are expected to also support non-blocking operation, which permits the `read` and `write` calls to return immediately when unable to transfer data. See reference of these functions for more information.
+
+For a timed wait for data, or to monitor multiple `FileHandle`s, `poll()` is provided in two forms:
+
+* `int poll(struct pollfd fds[], nfds_t nfds, int timeout)` standard POSIX form taking array of integer file descriptors
+* `int mbed::poll(pollfh fhs[], unsigned nfhs, int timeout)` variant taking an array of `FileHandle *`
+
+See the POSIX specification of `poll` for full details of the API. Mbed OS provides the `POLLIN`, `POLLOUT`, `POLLERR` and `POLLHUP` event flags, but implementation is dependent on device.
+
+As per POSIX, ordinary files always indicate readable and writable via poll.
+
+
+### Event-driven I/O
+
+If using non-blocking I/O, you will probably want to be know when to next attempt a `read` or `write` if they indicate no data is available. `FileHandle::sigio` lets you attach a `Callback` which will be called whenever the `FileHandle` becomes readable or writable.
+
+Important notes on sigio:
+
+* The sigio may be issued from interrupt context. You cannot portably issue `read` or `write` calls directly from this callback, so you should queue an `Event` or wake a thread to perform the `read` or `write`.
+* The sigio callback is only guaranteed when a `FileHandle` _becomes_ readable or writable. So if you do not fully drain the input, or fully fill the output, no sigio may be generated. This is also important on start-up - don't wait for sigio before attempting to read/write for the first time, but only use it as a "try again" signal after seeing an `EAGAIN` error.
+* Spurious sigios are permitted - you can't assume data will be available after a sigio.
+* Given all the above, use of sigio normally implies use of non-blocking mode, or possibly `poll`. 
+
+Ordinary files do not generate sigio callbacks, because they are always readable and writable, so never become readable or writable.
+
+### Stream-derived FileHandles
+
+Note that `FileHandle` implementations derived from `Stream`, such as `Serial`, have various limitations:
```

Comment:
This does kinda come out of no where. Should we describe the Stream class?
```suggestion
`Stream` is a legacy class that provides an abstract interface for streams similar to the `FileHandle` class. The difference is that the `Stream` API is built around the `getc` and `putc` set of functions, whereas `FileHandle` is build around `read` and `write`. This makes implementations simpler but limits what is possible with the API. Because of this, the FileHandle API is suggested API for new designs.

Note that `FileHandle` implementations derived from `Stream`, such as `Serial`, have various limitations:
```

Reply:
True. Will add an introductory paragraph along those lines.